worldedit-core/src/main/java/com/sk89q/worldedit/session/PlacementType.java

```diff
@@ -0,0 +1,110 @@
+import com.sk89q.worldedit.math.BlockVector3;
+import com.sk89q.worldedit.regions.RegionSelector;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+public enum PlacementType {
+    WORLD("worldedit.toggleplace.world", "worldedit.toggleplace.world-offset") {
+        @Override
+        public BlockVector3 getPlacementPosition(RegionSelector selector, Actor actor) throws IncompleteRegionException {
+            return BlockVector3.ZERO;
+        }
+    },
+
+    PLAYER("worldedit.toggleplace.player", "worldedit.toggleplace.player-offset") {
+        @Override
+        public BlockVector3 getPlacementPosition(RegionSelector selector, Actor actor) throws IncompleteRegionException
+        {
+            if (!canUseActor(actor)) {
+                throw new IncompleteRegionException();
+            }
+            return ((Locatable) actor).getBlockLocation().toVector().toBlockPoint();
+        }
+
+        @Override
+        public boolean canUseActor(Actor actor) {
+            checkNotNull(actor);
+            return actor instanceof Locatable;
+        }
+    },
+
+    HERE("worldedit.toggleplace.here", "worldedit.toggleplace.here-offset") {
+        @Override
+        public BlockVector3 getPlacementPosition(RegionSelector selector, Actor actor) throws IncompleteRegionException
+        {
+            throw new IllegalStateException("PlacementType.HERE.getPlacementPosition cannot be invoked. Use PLAYER or WORLD instead.");
+        }
+
+        @Override
+        public boolean canUseActor(Actor actor) {
+            return PLAYER.canUseActor(actor);
+        }
+    },
+
+    POS1("worldedit.toggleplace.pos1", "worldedit.toggleplace.pos1-offset") {
+        @Override
+        public BlockVector3 getPlacementPosition(RegionSelector selector, Actor actor) throws IncompleteRegionException {
+            return selector.getPrimaryPosition();
+        }
+    },
+
+    MIN("worldedit.toggleplace.min", "worldedit.toggleplace.min-offset") {
+        @Override
+        public BlockVector3 getPlacementPosition(RegionSelector selector, Actor actor) throws IncompleteRegionException {
+            return selector.getRegion().getMinimumPoint();
+        }
+    },
+
+    MAX("worldedit.toggleplace.max", "worldedit.toggleplace.max-offset") {
+        @Override
+        public BlockVector3 getPlacementPosition(RegionSelector selector, Actor actor) throws IncompleteRegionException {
+            return selector.getRegion().getMaximumPoint();
+        }
+    };
+
+    private final String translationKey;
+    private final String translationKeyWithOffset;
+
+    PlacementType(String translationKey, String translationKeyWithOffset) {
+        this.translationKey = translationKey;
+        this.translationKeyWithOffset = translationKeyWithOffset;
+    }
+
+    public abstract BlockVector3 getPlacementPosition(RegionSelector selector, Actor actor) throws IncompleteRegionException;
+
+    public boolean canUseActor(Actor actor) {
```

Comment:
feels like it would make more sense if this (and its derivatives) was named `canActorUse` or more verbosely `canBeUsedBy(Actor)` or something else that makes the subject and object more obvious.

Reply:
will be fixed in my next push